modulus/models/module.py

```diff
@@ -65,55 +106,286 @@ def debug(self):
+
+        # Define the signature of new init
+        params = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)]
+        params += [
+            inspect.Parameter(
+                argname,
+                inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                default=defaults_dict.get(argname, inspect.Parameter.empty),
+            )
+            for argname in model_argnames
+        ]
+        init_signature = inspect.Signature(params)
+
+        # Replace ModulusModel.__init__ signature with new init signature
+        ModulusModel.__init__.__signature__ = init_signature
+
+        # Generate a unique name for the created class
+        new_class_name = f"{torch_model_class.__name__}ModulusModel"
+        ModulusModel.__name__ = new_class_name
+
+        # Add this class to the dict of models classes
+        ModulusModel.register(ModulusModel, new_class_name)
+
+        return ModulusModel
+
+    @classmethod
+    def register(cls, model, name: Union[str, None] = None):
+        """
+        Registers a model under a specific name.
+
+        Parameters
+        ----------
+        model
+            The model to be registered.
+        name : Union[str,None], optional
+            The name to register the model under. If none is provided, the model's name
+        """
+
+        # If no name provided, use the model's name
+        if name is None:
+            name = model.__name__
+
+        # Check if name already in use
+        if name in Module._model_registry:
+            raise ValueError(f"Name {name} already in use")
+
+        # Add this class to the dict of model registry
+        Module._model_registry[name] = model
+
+    @classmethod
+    def factory(cls, name: str):
+        """
+        Returns a registered model given its name.
+
+        Parameters
+        ----------
+        name : str
+            The name of the registered model.
+
+        Returns
+        -------
+        model
+            The registered model.
+
+        Raises
+        ------
+        KeyError
+            If no model is registered under the provided name.
+        """
+
+        if name in Module._model_registry:
+            model = Module._model_registry[name]
+            if isinstance(model, pkg_resources.EntryPoint):
+                model = model.load()
+            return model
+        else:
+            raise KeyError(f"No model is registered under the name {name}")
+
+    @classmethod
+    def list_models(cls) -> List[str]:
+        """
+        Returns a list of registered model names.
+
+        Returns
+        -------
+        List[str]
+            List of registered model names.
+        """
+
+        return list(Module._model_registry.keys())
+
+    @classmethod
+    def _clear_model_registry(cls):
+        # NOTE: This is only used for testing purposes
+        Module._model_registry = {}
+
+    @classmethod
+    def _restore_model_registry(cls):
+        # NOTE: This is only used for testing purposes
```

Comment:
Should name this `__restore_model_registry` with two underscores to indicate fully private use only